clisops/core/regrid.py

```diff
@@ -0,0 +1,1854 @@
+        except (KeyError, AttributeError):
+            warnings.warn(
+                "For coordinate variable '%s' no bounds can be identified." % coordinate
+            )
+            return
+
+    def _grid_detect_collapsed_cells(self):
+        "Detect collapsing grid cells. Requires defined bounds."
+        mask_lat = self._create_collapse_mask(self.ds[self.lat_bnds].data)
+        mask_lon = self._create_collapse_mask(self.ds[self.lon_bnds].data)
+        # for regular lat-lon grids, create 2D coordinate arrays
+        if (
+            mask_lat.shape != mask_lon.shape
+            and mask_lat.ndim == 1
+            and mask_lon.ndim == 1
+        ):
+            mask_lon, mask_lat = np.meshgrid(mask_lon, mask_lat)
+        self.coll_mask = mask_lat | mask_lon
+        self.contains_collapsed_cells = bool(np.any(self.coll_mask))
+
+    def _create_collapse_mask(self, arr):
+        "Grid cells collapsing to lines or points"
+        orig_shape = arr.shape[:-1]  # [nlon, nlat, nbnds] -> [nlon, nlat]
+        arr_flat = arr.reshape(-1, arr.shape[-1])  # -> [nlon x nlat, nbnds]
+        arr_set = np.apply_along_axis(lambda x: len(set(x)), -1, arr_flat)
+        mask = np.zeros(arr_flat.shape[:-1], dtype=bool)
+        mask[arr_set == 1] = True
+        return mask.reshape(orig_shape)
+
+    def _cap_precision(self, decimals):
+        """
+        Round horizontal coordinate variables to specified precision using numpy.around.
+
+        Parameters
+        ----------
+        decimals : int
+            The decimal position / precision to round to.
+
+        Returns
+        -------
+        None.
+        """
+        # todo: extend for vertical axis for vertical interpolation usecase
+        # 6 decimals corresponds to hor. precision of ~ 0.1m (deg), 6m (rad)
+        coord_dict = {}
+        attr_dict = {}
+        encoding_dict = {}
+
+        # Assign the rounded values as new coordinate variables
+        for coord in [self.lat_bnds, self.lon_bnds, self.lat, self.lon]:
+            if coord:
+                attr_dict.update({coord: self.ds[coord].attrs})
+                encoding_dict.update({coord: self.ds[coord].encoding})
+                coord_dict.update(
+                    {
+                        coord: (
+                            self.ds[coord].dims,
+                            np.around(self.ds[coord].data.astype(np.float64), decimals),
+                        )
+                    }
+                )
+
+        # Restore the original attributes
+        if coord_dict:
+            self.ds = self.ds.assign_coords(coord_dict)
+            # Restore attrs and encoding - is there a proper way to do this?? (todo)
+            for coord in [self.lat_bnds, self.lon_bnds, self.lat, self.lon]:
+                if coord:
+                    self.ds[coord].attrs = attr_dict[coord]
+                    self.ds[coord].encoding = encoding_dict[coord]
+
+    def _compute_hash(self):
```

Comment:
I think this could be useful for xESMF as well. I suggest we port this function upstream.